gateway-service/src/main/java/org/zowe/apiml/gateway/routing/ZuulErrorControllerHandler.java

```diff
@@ -0,0 +1,83 @@
+import org.springframework.beans.factory.config.BeanPostProcessor;
+import org.springframework.cglib.proxy.*;
+import org.springframework.cloud.netflix.zuul.filters.RouteLocator;
+import org.springframework.cloud.netflix.zuul.web.ZuulController;
+import org.springframework.cloud.netflix.zuul.web.ZuulHandlerMapping;
+import org.springframework.stereotype.Component;
+import org.zowe.apiml.product.compatibility.ApimlErrorController;
+
+import java.lang.reflect.Constructor;
+import java.lang.reflect.Method;
+
+/**
+ * This class is used to reconcile the breaking change between Spring Boot 2.5 and Zuul. The breaking change
+ * is due to ErrorController.getErrorPath being removed in Spring Boot 2.5. A BeanPostProcessor is used
+ * to proxy ZuulHandlerMapping, intercepting the code execution that leads to the NoSuchMethodError.
+ * <p>
+ * NOTE: This should be removed when the APIML migrates away from Zuul to Spring Cloud Gateway.
+ */
+@Component
+public class ZuulErrorControllerHandler implements BeanPostProcessor {
+    private final RouteLocator routeLocator;
+    private final ZuulController zuulController;
+    private final ApimlErrorController errorController;
+
+    public ZuulErrorControllerHandler(RouteLocator routeLocator, ZuulController zuulController, ApimlErrorController errorController) {
+        this.routeLocator = routeLocator;
+        this.zuulController = zuulController;
+        this.errorController = errorController;
+    }
+
+    @Override
+    public Object postProcessAfterInitialization(Object bean, String beanName) throws BeansException {
+        if (errorController != null && (bean instanceof ZuulHandlerMapping)) {
+            Enhancer enhancer = new Enhancer();
+            enhancer.setSuperclass(ZuulHandlerMapping.class);
+            enhancer.setCallbackFilter(LookupHandlerCallbackFilter.INSTANCE);
+            enhancer.setCallbacks(new Callback[]{new LookupHandlerMethodInterceptor(errorController.getErrorPath()), NoOp.INSTANCE});
+            Constructor<?> ctor = ZuulHandlerMapping.class.getConstructors()[0];
+            return enhancer.create(ctor.getParameterTypes(), new Object[]{routeLocator, zuulController});
+        }
+        return bean;
+    }
+
+    private enum LookupHandlerCallbackFilter implements CallbackFilter {
+        INSTANCE;
+
+        @Override
+        public int accept(Method method) {
+            if ("lookupHandler".equals(method.getName())) {
+                return 0;
+            }
+            return 1;
+        }
+    }
+
+    private static final class LookupHandlerMethodInterceptor implements MethodInterceptor {
+        private final String errorPath;
+
+        LookupHandlerMethodInterceptor(String errorPath) {
+            this.errorPath = errorPath;
+        }
+
+        @Override
+        public Object intercept(Object target, Method method, Object[] args, MethodProxy methodProxy) throws Throwable {
+            if (errorPath.equals(args[0])) {
+                return null;
```

Comment:
I don't understand the use of this code. It works only for 1 path, even though there are multiple implementations of ApimlErrorController. Even without this block of code, the client gets the same response. What is the reason of returning null only for "internal_error" url? Why it ignores others, like "not_found"?

Reply:
Previously, all handling of error paths were handled checking against `InternalServerErrorController` only, because we use `ZuulServerAutoConfiguration` which calls `ZuulHandlerMapping.setErrorController` using `InternalServerErrorController`. Now, the enhanced `ZuulHandlerMapping` bean only ever checks error paths against a `null` `ErrorController` instance, bypasses the `NoSuchMethodError`. So the behaviour of what `ErrorController` is used hasn't changed, just where it is used has changed to avoid the error.

That said, I agree this is strange and unintuitive, so I've pushed a commit that checks the error path for all `ApimlErrorController` instances before calling `lookUpHandler`.